Draft: Add security handbook (#115)
lf-lang/lf-lang.github.io: packages/documentation/copy/en/reference/Security.md

---
title: "Security"
layout: docs
permalink: /docs/handbook/security
oneline: "Secure Federated Execution"
preamble: >
---

> ### Warning: This part is under construction. Users should not expect federations are secure applying this property.

Reviewer: ```suggestion
### Warning: the experimental security features described on this page are under development and not production ready. Users should not expect their federations will be secure if the `auth` target property is enabled.
```

> By default, there is no secure authentication happening when a federate joins a federation, and data exchanged by federates is not encrypted. The `auth` target property can be used to change this.

Reviewer: ```suggestion
By default, there is no secure authentication happening when a federate joins a federation, and data exchanged by federates is not encrypted. For targets that support it, the `auth` target property can be used to enable authentication between federates. Messages exchanged between federates after authentication are **not encrypted**, but this capability is planned for the future.
```

---
title: "Security"
layout: docs
permalink: /docs/handbook/security
oneline: "Federated Execution with Security"
preamble: >
---

> The current implementation of federated execution lacks authentication of federates from joining the federation. Also communication security is not implemented to prevent adversarial federates sending malicious messages (e.g.. bad tags or bad sensor readings) to the RTI or other federates.

Reviewer: ```suggestion
By default, there is no secure authentication happening when a federate joins a federation, and data exchanged by federates is not encrypted. The `auth` target property can be used to change this.
```

Author: Commited all your changes at once including this and below.

<div class="lf-cpp lf-py lf-ts lf-rs">

The $target-language$ target does not currently support the `auth` target option.

</div>

<div class="lf-c">

## Authentication

> For the `C` target at least, federated execution is able to apply security with authentication by using HMAC authentication between RTI and federates. To enable this, include the `auth` property in your target specification, as follows:

Reviewer: ```suggestion
For the `C` target, federated execution is able to apply security with authentication by using HMAC authentication between RTI and federates. To enable this, include the `auth` property in your target specification, as follows:
```

```lf-c
target C {
    auth: true
};
```

> The RTI build must include cmake options to enable simple HMAC-based authentication of federates. Add `-DAUTH=ON` option to the cmake command as shown below:

Reviewer: ```suggestion
The RTI build must include CMake options to enable simple HMAC-based authentication of federates. Add `-DAUTH=ON` option to the CMake command as shown below:
```

```bash
mkdir build && cd build
cmake -DAUTH=ON ../
make
sudo make install
```

If you would like to go back to non-AUTH mode, you would have to remove all contents of the `build` folder.

## Future Work

In the future, we will be able to use full-fledged authentication, for example, using different keys for each federate or using public-key cryptography for authentication. Also, we can try to apply different encryption and message authentication algorithms for each federate in the same federation, to support a federation with federates running on heterogeneous platforms where some of platforms have limited resources or capabilities.

### Communication Security

> Communication security is to ensure confidentiality, integrity and message authenticity of the network messages between RTI and federates as well as among federates.

Reviewer: It is unclear how this text relates to the target property. Isn't encryption implied by the use of the target property?

Reviewer: What about this part? This text seems out of place...

Author: I added more details of this part. The encryption is not applied to the messages for communication yet, and this part is the description of the plan.

Reviewer: Hm, this sounds like a very experimental feature then. Should we even be documenting it if it is so incomplete? I suggest we put a fat warning message at the very beginning of the page about this so that users don't enable the feature with the expectation their federation will be secure...

Author: Yeah. I was actually starting to doubt myself 'Does the handbook need to show future works.' 
Then I'll erase the future work part and add a warning message.

Reviewer: I think we can safely leave out the details for future work. Those would be good to document in an issue, however.

The current implementation of federated execution lacks authentication of federates from joining the federation. Also communication security is not implemented to prevent adversarial federates sending malicious messages (e.g.. bad tags or bad sensor readings) to the RTI or other federates.

<div class="lf-cpp lf-py lf-ts lf-rs">

The $target-language$ target does not currently support the `auth` target option.

</div>

<div class="lf-c">

## Authentication

For the `C` target at least, federated execution is able to apply security with authentication by using HMAC authentication between RTI and federates. To enable this, include the `auth` property in your target specification, as follows:

```lf-c
target C {
    auth: true
};
```

> The RTI bui0ld must include cmake options to enable simple HMAC-based authentication of federates. Add `-DAUTH=ON` option to the cmake command as shown below:

Reviewer 2: ```suggestion
The RTI build must include cmake options to enable simple HMAC-based authentication of federates. Add `-DAUTH=ON` option to the cmake command as shown below:
```

Author: @edwardalee Thanks for checking my PR. I fixed what you mentioned on the latest commit 7972f11 . Is there any other suggestions to add or change?
